### PR TITLE
[DRAFT] offsets for free

### PR DESCRIPTION
**Pull request by Ngoguey42**
This is the result of my attempt to implement structured key's optimisation regarding calls to `index.find` for offsets that were known during the inode deserialisation.

This is overall not an improvement, I'm posting it here just in case someone is interested.

---

I implemented it using an LRU, the size of the LRU doesn't seem to matter a lot. I chose 1000. 1million showed similar performences.

Before the LRU I tried using ephemerons but it failed mysteriously because of the size of the weak map that reached tens of millions. I was expecting that during the fold we would only have ~1000 inodes in memory (I'm not refering to the new inodes here, only the ones read from disk in order to perfom the fold).

Some stats look good,
```
                                           |              |      total      
Disk bytes read                            | small        |  192.381 G      
                                           | small        |  192.381 G 100% 
                                           | small_offlru |  122.500 G  64% 
Disk reads                                 | small        |  241.853 M      
                                           | small        |  241.853 M 100% 
                                           | small_offlru |  150.927 M  62% 
pack.cache_misses                          | small        |   37.420 M      
                                           | small        |   37.420 M 100% 
                                           | small_offlru |    3.941 M  11% 
gc.minor_words allocated                   | small        |   81.775 G      
                                           | small        |   81.775 G 100% 
                                           | small_offlru |   68.981 G  84% 
rusage.stime                               | small        |    1min40s      
                                           | small        |    1min44s 104% 
                                           | small_offlru |     1min1s  61% 

```

and some don't:
```
gc.major_words allocated                   | small        |    5.334 G      
                                           | small        |    5.334 G 100% 
                                           | small_offlru |   12.957 G 243% 
rusage.maxrss                              | small        |    6.397 G      
                                           | small        |    6.397 G 100% 
                                           | small_offlru |    9.427 G 147% 
rusage.utime                               | small        |    6min27s      
                                           | small        |    6min30s 101% 
                                           | small_offlru |    7min59s 124% 
```

The new counters I added
```
"offlru_insertions":       70_322_991,

"unsafe_find":             41_751_735,
"unsafe_find_table_hit":            0,
"unsafe_find_vallru_hit":   4_335_565,
"unsafe_find_offlru_hit":  33_478_922, <============
"unsafe_find_long_none":            0,
"unsafe_find_long_some":    3_937_248,
 
"unsafe_append":                     2_605_446,
"unsafe_append_offset":             12_416_722,
"unsafe_append_offset_offlru_hit":   3_921_792, <============
"unsafe_append_offset_long_none":            0,
"unsafe_append_offset_long_some":    8_494_930,

"blindfolded_read_first_try":       33_380_443,
"blindfolded_read_second_try":          98_479,
"blindfolded_read_third_or_more_try":        0,
```

### CODE REVIEW

**Comment by Ngoguey42** [comment]
Using `cachecache` instead of `Lru` doesn't change much the stats. It represents a small improvement in major words allocated

```
                         |              |      total      
                         |              |                 
gc.minor_words allocated | small        |   81.775 G      
                         | small_offlru |   68.981 G  84% 
                         | small_offcc  |   68.506 G  84% 
gc.major_words allocated | small        |    5.334 G      
                         | small_offlru |   12.957 G 243% 
                         | small_offcc  |   12.563 G 236% 
```

**Comment by Ngoguey42** [comment]
TLDR; we need shared buffer for reads everywhere in `pack_store.ml`

---

This PR now uses a shared buffer in `pack_store.ml` that is used for "blindfolded" reads. Replaying 100k blocks shows similar performances with `main`.

The `Disk bytes read` is very similar between `main` and this PR, even though blindfolded reads read full pages. It indicates that what is gained on reading index is spent on reading pack. In other words, it removes reads from index and the pack reads that were going to happen anyway are larger and aligned on pages.

For both `main` and this PR a shared buffer could also be used for "normal" reads. I will open another PR that just introduces that shared buffer for `main`.

The full stats: [full.txt](https://github.com/mirage/irmin/files/7561540/full.txt)

A subset of the stats:
```
Block played count *C                  |      | 0 (before) |     12500      |      25000      |      37500      |      50000      |      62500      |      75000      |      87500      |  100000 (end)
Blocks progress *C                     |      |     0%     |      12%       |       25%       |       38%       |       50%       |       62%       |       75%       |       88%       |      100%

Wall time elapsed *C                   | main |        0us |    1min6s      |    1min46s      |     3min3s      |    4min52s      |       7min      |    9min15s      |   11min48s      |   14min28s
                                       | main |        0us |    1min7s 102% |    1min48s 101% |     3min3s 100% |    4min53s 100% |     7min2s 101% |    9min19s 101% |   11min49s 100% |   14min25s 100%
                                       | off  |        0us |   1min12s 109% |    1min55s 108% |    3min16s 107% |     5min7s 105% |    7min18s 104% |    9min37s 104% |    12min8s 103% |   14min51s 103%
                                       | off  |        0us |    1min7s 102% |    1min48s 101% |     3min8s 103% |     5min2s 103% |    7min19s 105% |    9min38s 104% |   12min15s 104% |   14min53s 103%

gc.minor_words allocated *C            | main |    0.001 G |  10.065 G      |   16.420 G      |   27.023 G      |   41.152 G      |   57.617 G      |   75.395 G      |   95.013 G      |  115.837 G
                                       | main |    0.001 G |  10.065 G 100% |   16.420 G 100% |   27.025 G 100% |   41.155 G 100% |   57.625 G 100% |   75.315 G 100% |   94.989 G 100% |  115.773 G 100%
                                       | off  |    0.001 G |  10.076 G 100% |   16.437 G 100% |   27.038 G 100% |   41.160 G 100% |   57.577 G 100% |   75.027 G 100% |   94.618 G 100% |  115.347 G 100%
                                       | off  |    0.001 G |  10.076 G 100% |   16.437 G 100% |   27.038 G 100% |   41.169 G 100% |   57.611 G 100% |   75.275 G 100% |   94.915 G 100% |  115.698 G 100%

gc.major_words allocated per block *LA | main |        n/a |    56_332      |     31_981      |     64_512      |     80_641      |     88_379      |     94_126      |    100_451      |    111_325
                                       | main |        n/a |    56_222 100% |     32_100 100% |     64_800 100% |     80_647 100% |     88_255 100% |     92_999  99% |    100_925 100% |    110_799 100%
                                       | off  |        n/a |    56_253 100% |     31_892 100% |     65_192 101% |     80_782 100% |     87_921  99% |     92_678  98% |    100_382 100% |    111_436 100%
                                       | off  |        n/a |    56_740 101% |     32_350 101% |     65_273 101% |     80_728 100% |     88_689 100% |     93_342  99% |    100_925 100% |    111_368 100%

Disk bytes read    per block *LA       | main |        n/a |    25_961      |     13_233      |    801_616      |  1_027_384      |  1_166_420      |  1_044_205      |  1_261_782      |  1_017_340
                                       | main |        n/a |    25_961 100% |     13_233 100% |    802_063 100% |  1_027_667 100% |  1_167_799 100% |  1_000_677  96% |  1_271_347 101% |  1_015_044 100%
                                       | off  |        n/a |    48_801 188% |     28_108 212% |    780_232  97% |  1_007_593  98% |  1_131_008  97% |    967_065  93% |  1_267_108 100% |  1_012_993 100%
                                       | off  |        n/a |    48_801 188% |     28_108 212% |    779_965  97% |  1_010_070  98% |  1_139_832  98% |    986_228  94% |  1_247_856  99% |  1_009_240  99%

pack.cache_misses per block *LA        | main |        n/a |    30.407      |     16.037      |     44.331      |     46.467      |     50.092      |     56.039      |     57.046      |     69.114
                                       | main |        n/a |    30.407 100% |     16.037 100% |     44.331 100% |     46.467 100% |     50.092 100% |     56.039 100% |     57.046 100% |     69.114 100%
                                       | off  |        n/a |    18.637  61% |      8.351  52% |     29.196  66% |     31.289  67% |     33.706  67% |     38.357  68% |     38.535  68% |     47.003  68%
                                       | off  |        n/a |    18.637  61% |      8.351  52% |     29.196  66% |     31.289  67% |     33.706  67% |     38.357  68% |     38.535  68% |     47.003  68%

 *LA: Local Average. Smoothed using a weighted sum of the value in the
      block and the exponentially decayed values of the previous blocks.
      Every 1250.01 blocks, half of the past is forgotten.
```

**Comment by Ngoguey42** [comment]
It seems impossible to witness the perfs improvements of the LRU during the bootstrap replay. The LRU hits only ~2k times per second while it hit ~70k times per second during the migration.

**Comment by Ngoguey42** [comment]
While this PR makes no difference on the block replay benchmark, it makes a difference on the synthetic migration benchmark:
```
                                          minor          major  utime  stime     wall  maxrss  minor-page-fault
main                              7_592_023_127    503_889_117  61.56  29.21  1:30.78  231404  119085
main                              7_592_023_127    503_889_117  61.80  29.39  1:31.20  231276  112354
offsets for free                  5_700_616_159  1_870_406_690  57.73   9.28  1:07.02  297252   83571
offsets for free                  5_700_616_159  1_870_406_690  59.73   9.79  1:09.52  297164  135672
offsets for free + shared buffer  5_264_561_692    456_971_361  36.20   8.68  0:44.88  263268  127027
offsets for free + shared buffer  5_264_561_692    456_971_361  36.48   8.27  0:44.76  263280  149706
```

That's a 50% improvements in wall time

**Comment by samoht** [comment]
> That's a 50% improvements in wall time

That's super impressive!

**Comment by Ngoguey42** [comment]
`main ` vs `offsets for free` showed very similar results on the real migration (see the first messages in this PR). 

I suspect that `main ` vs `offsets for free + shared buffer` would be different but I don't have the setup to test that anymore.

The new LRU in `unsafe_find` hits at 90% during the migration, the structured keys will represent a 99% "hit" rate, which will represent an even bigger improvement in the "migration" access pattern setup. Too bad it will arrive too late for the upcoming migration :)

**Comment by Ngoguey42** [comment]
Closing as this is not such a big improvement
